### DISCUSSION TITLE
Invoking Event for different users

### DISCUSSION THREAD

**Discussion by jnsw** [eclipse-milo/milo#1036] [General]
Is it somewhere possible to invoke events, but only the admin can get "all" properties of that event? And the customer will get the same event, but two properties should be missing. I made two distinct events, made one of them invisible to the customer, but if I am invoking both, the customer will also get both even he can't see the BaseEventType-extended event. Is there already a solution inside, I could use? I made it invisible with your solution in https://github.com/eclipse/milo/discussions/922#discussioncomment-2790754 but it can be still seen in the UaExpert Event View as subscribable event

**Comment by kevinherron** [comment]
I don't think OPC UA permission model has explicit support for this, so right now my only idea would be to create two separate Objects that will act as Event Notifiers, one for "user", and one for "admin", and make sure that when a "user" reads the EventNotifier attribute of the "admin" Object the SubscribeToEvents bit is not set.

Then you would publish your separate user/admin events via these separate EventNotifier Objects.

**Reply by kevinherron** [reply]
One problem with the above is that it breaks the expectation that the Server object is able to be used as the Notifier that notifies of _all_ events in the server, sitting at the top of a hierarchy of Notifiers.

**Reply by kevinherron** [reply]
Okay, so the OPC UA 1.03 permission model doesn't have any way to deal with this, but 1.04+ does (though it's complicated and requires the server have roles and permissions system set up), so maybe in the future there will be a "first class" way to support this in the SDK, but for now you have to use the tacky two Notifier solution.

**Reply by jnsw** [reply]
yeah, but the tacky two notifier solution gets both events also to the customer, can I restrict even subscribe to it?

**Reply by kevinherron** [reply]
Yes, by making it so when a regular "user" reads the EventNotifier attribute it doesn't indicate the SubscribeToEvents bit it set.

You'll need to install an AttributeFilter that looks at the incoming user or something. Similar to how this filter works in the demo server: https://github.com/digitalpetri/opc-ua-demo-server/blob/master/src/main/kotlin/com/digitalpetri/opcua/server/namespaces/filters/ExecutableByAdminFilter.kt

**Reply by jnsw** [reply]
wait, how can I specify not to subscribe to one of my created EventTypes, that are derived from BaseEventType? with this solution I would revoke him to subscribe at all, am I?

**Reply by kevinherron** [reply]
You don't mark certain event types as not allowed to receive, you mark an entire EventNotifier Object as not subscribable to non-admin users. Then you only send admin events through the admin EventNotifier.

**Reply by jnsw** [reply]
how do i specify in invoking an event the other EventNotifier? like in https://github.com/eclipse/milo/blob/3419235853491ac89f3391bffd337c8200522522/milo-examples/server-examples/src/main/java/org/eclipse/milo/examples/server/ExampleNamespace.java#L223 i can't see specifying any EventNotifier 😅

**Reply by kevinherron** [reply]
An Event Notifier is just an Object that will act as an event source, and indicates it can be an event source by setting the SubscribeToEvents bit in its EventNotifier attribute.

By default the Server Object acts as an Event Notifier. You're going to need to create an additional Object that will act as Event Notifier, but only for "admin" users, and you'll do that by controlling the value that gets returned for its EventNotifier attribute based on the user that is reading it. You'll need another EventBus-like object to send events to, and your Namespace/AddressSpace will need handling for events and this EventBus [similar to OpcUaNamespace](https://github.com/eclipse/milo/blob/3419235853491ac89f3391bffd337c8200522522/opc-ua-sdk/sdk-server/src/main/java/org/eclipse/milo/opcua/sdk/server/namespaces/OpcUaNamespace.java#L124-L145).
